okhttp/src/main/java/okhttp3/ConnectionSpec.kt

```diff
@@ -0,0 +1,324 @@
+    val specToApply = supportedSpec(sslSocket, isFallback)
+
+    if (specToApply.tlsVersions != null) {
+      sslSocket.enabledProtocols = specToApply.tlsVersions
+    }
+
+    if (specToApply.cipherSuites != null) {
+      sslSocket.enabledCipherSuites = specToApply.cipherSuites
+    }
+  }
+
+  /**
+   * Returns a copy of this that omits cipher suites and TLS versions not enabled by `sslSocket`.
+   */
+  private fun supportedSpec(sslSocket: SSLSocket, isFallback: Boolean): ConnectionSpec {
+    var cipherSuitesIntersection = if (cipherSuites != null) {
+      intersect(CipherSuite.ORDER_BY_NAME, sslSocket.enabledCipherSuites, cipherSuites)
+    } else {
+      sslSocket.enabledCipherSuites
+    }
+
+    val tlsVersionsIntersection = if (tlsVersions != null) {
+      intersect(Util.NATURAL_ORDER, sslSocket.enabledProtocols, tlsVersions)
+    } else {
+      sslSocket.enabledProtocols
+    }
+
+    // In accordance with https://tools.ietf.org/html/draft-ietf-tls-downgrade-scsv-00 the SCSV
+    // cipher is added to signal that a protocol fallback has taken place.
+    val supportedCipherSuites = sslSocket.supportedCipherSuites
+    val indexOfFallbackScsv = indexOf(
+        CipherSuite.ORDER_BY_NAME, supportedCipherSuites, "TLS_FALLBACK_SCSV")
+    if (isFallback && indexOfFallbackScsv != -1) {
+      cipherSuitesIntersection = concat(
+          cipherSuitesIntersection, supportedCipherSuites[indexOfFallbackScsv])
+    }
+
+    return Builder(this)
+        .cipherSuites(*cipherSuitesIntersection)
+        .tlsVersions(*tlsVersionsIntersection)
+        .build()
+  }
+
+  /**
+   * Returns `true` if the socket, as currently configured, supports this connection spec. In
+   * order for a socket to be compatible the enabled cipher suites and protocols must intersect.
+   *
+   * For cipher suites, at least one of the [required cipher suites][cipherSuites] must match the
+   * socket's enabled cipher suites. If there are no required cipher suites the socket must have at
+   * least one cipher suite enabled.
+   *
+   * For protocols, at least one of the [required protocols][tlsVersions] must match the socket's
+   * enabled protocols.
+   */
+  fun isCompatible(socket: SSLSocket): Boolean {
+    if (!isTls) {
+      return false
+    }
+
+    if (tlsVersions != null
+        && !nonEmptyIntersection(Util.NATURAL_ORDER, tlsVersions, socket.enabledProtocols)) {
+      return false
+    }
+
+    if (cipherSuites != null
+        && !nonEmptyIntersection(
+            CipherSuite.ORDER_BY_NAME, cipherSuites, socket.enabledCipherSuites)) {
+      return false
+    }
+
+    return true
+  }
+
+  override fun equals(other: Any?): Boolean {
+    if (other !is ConnectionSpec) return false
+    if (other === this) return true
+
+    val that = other as ConnectionSpec?
+    if (this.isTls != that!!.isTls) return false
+
+    if (isTls) {
+      if (!Arrays.equals(this.cipherSuites, that.cipherSuites)) return false
```

Comment:
Kotlin uses `contentEquals` extension on array. It just ends up calling this though so not a big deal

Reply:
Tried it, and ran into a type inference issue. Will come back to it later.

Reply:
(I think I’ll be happier migrating this class to use Lists instead of Arrays.)